stdlib/asm/crypto/dsa/falcon.masm

```diff
@@ -1,278 +1,611 @@
 #!
-#! If verification fails, program panics, due to failure in assertion !
+#! Input: [PK, MSG, NONCE1, NONCE1, ...]
+#! Output: [...]
 #!
-#! Note, input memory addresses are considered to be immutable.
-export.verify.257
-    locaddr.0
-    movdn.2
-    exec.poly512::mul_zq
+#! Cycles: ~ 93363
+export.verify.1665
 
-    locaddr.128
-    locaddr.0
-    exec.poly512::neg_zq
+    # 1) Load the public key polynomial h and the signature polynomial s2 and the product of
+    # the two polynomials pi := h * s2 in Z_Q[x]. This also checks that h hashes to the provided
+    # digest PK. While loading the polynomials, the hash of the three polynomials is computed
+    # and the first half of the digest is kept on the stack for later use by the
+    # `probablistic_product` procedure.
 
     locaddr.0
-    swap
-    locaddr.128
-    exec.poly512::add_zq
+    exec.load_h_s2_and_product
+    #=> [tau1, tau0, tau_ptr, MSG, NONCE1, NONCE1, ...]      (Cycles: 571)
 
-    locaddr.128
-    locaddr.0
-    exec.normalize_poly512
+    exec.powers_of_tau
+    #=> [zeros_ptr, MSG, NONCE1, NONCE1, ...]                (Cycles: 8323)
 
-    # compute squared norm of s0
+    exec.set_to_zero
+    #=> [c_ptr, MSG, NONCE1, NONCE1, ...]                    (Cycles: 2607)
 
-    locaddr.128
-    exec.squared_norm_poly512
+    # 2) Compute the hash-to-point of the message MSG from the provided NONCE and save
+    # the resulting polynomial c in the memory region [c_ptr, c_ptr + 128)
 
-    locaddr.256
-    mem_store
+    exec.hash_to_point
+    #=> [...]                                       (Cycles: 1884)
 
-    # compute squared norm of s1 ( where s1 is provided as polynomial
-    # with coefficients represented using absolute value i.e. signs are ignored )
+    # 3) Check that we indeed have pi := h * s2 in Z_Q[x] by checking that pi(tau) = h(tau) * s2(tau)
+    # where tau is a random (Fiat-Shamir) challenge resulting from hashing h, s2 and pi.
 
-    exec.squared_norm_poly512
+    locaddr.512     # tau_ptr
+    locaddr.1025    # z_ptr
+    locaddr.0       # h ptr
+    #=> [h_ptr, zeros_ptr, tau_ptr, ...]
 
+    exec.probablistic_product
+    #=> [...]                                       (Cycles: 7760)
+
+    # 4) Compute the squared norm of s1 := c - h * s2 (in Z_q[x]/(phi))
+    
     locaddr.256
-    mem_load
-    add
+    #=> [pi_ptr, ...]
 
-    # check that norm of the signature is small enough
+    exec.compute_s1_norm_sq
+    #=> [norm_sq(s1), ...]                          (Cycles: 58888)  
 
-    push.34034726 # constant sig_bound for Falcon-512 signature
-    lte
-    assert
-end
+    # 5) Compute the squared norm of s2
+
+    locaddr.128
+    #=> [s2_ptr, norm_sq(s1), ...]                  
+
+    exec.compute_s2_norm_sq                         
+    #=> [norm_sq(s2), norm_sq(s1), ...]             (Cycles: 13322)
+
+    # 6) Check that ||(s1, s2)||^2 < K
+
+    add
+    #=> [norm_sq(s1) + norm_sq(s2), ...]
+
+    push.34034726
```

Comment:
nit: would be good to replace this value with a constant.

Reply:
Done!